### PR TITLE
dev/core#1046 - Fix name vs label for case roles

### PR DESCRIPTION
**Pull request by demeritcowboy**
Overview
----------------------------------------
There is a longstanding issue where name and label got mixed at some point for case roles which then comes up in multiple scenarios. This fixes at least the following:

* Fresh out of the box, changing the default language and then enabling civicase causes fatal errors with `drush cc civicrm`, enabling extensions, and creating a new case. [dev/core#329](https://lab.civicrm.org/dev/core/issues/329)
* Also fresh out of the box, enabling civicase and changing the label of a relationship type causes various problems, including a fatal error if it's the case manager role. This is another way to cause the above fatal errors since it's the same root cause. [dev/core#774](https://lab.civicrm.org/dev/core/issues/774) and [dev/core#856](https://lab.civicrm.org/dev/core/issues/856)
* People using xml files need to update their xml files if they change the label of a relationship type (defeating the purpose of `name`).

I can break this PR down into some smaller chunks but I wanted to put up the whole thing because in order to try it out you need the whole thing. The meat of it is the 3 files under CRM/Case and ang/crmCaseType.js, which I don't think I can break up too much because a change in one place throws everything else off.

It also adds some system status checks to warn about edge cases, e.g. if you have somehow ended up with a label that matches another type's machine name, or roles you have in your xml file that don't exist as relationship types. There is a WIP addition to the user guide [here](https://github.com/civicrm/civicrm-user-guide/compare/master...demeritcowboy:case-role-status-checks-help) that goes along with this. I'm not sure it belongs in the user guide but that's where the other case setup/admin docs are.

Before
----------------------------------------
Case roles broken if you change a label.

After
----------------------------------------
Better

Technical Details
----------------------------------------
It is not a perfect fix but I haven't been able to break it in testing so am satisfied for the moment. I have some more thoughts which are related but maybe out of scope so I'll collect them and post in the lab ticket.

Comments
----------------------------------------
@colemanw @jitendrapurohit @alifrumin @agh1 @pradpnayak @bgm @ErikHommel @varshith @mattwire @totten

### CODE REVIEW

**Comment by demeritcowboy** [comment]
api4 test seems fine locally - jenkins retest this please

**Comment by alifrumin** [comment]
@demeritcowboy I can review this this week

**Comment by demeritcowboy** [comment]
Thanks @alifrumin !

**Review by alifrumin — COMMENTED** [review]
@agh1 and I reviewed this together, generally we think that this moves things forward in a positive way. It is well documented, adds system checks, improves test coverage and fixes some gnarly bugs. We found one related bug that we think could easily be fixed as a part of this pr, had some suggestions of how to improve user experience by using different PSR severity levels/ icons and some suggestions of how to improve readability of the code. See summary below, the most detailed notes are in the code review itself.

* General standards
    * Explain ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) -  __PASS__ : The goal/problem/solution has been adequately explained in the PR and related gitlab issue.
    * User impact ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) - __PASS__: The changes would be intuitive or unnoticeable for a majority of users who work with this feature. For the new System Status checks we added some suggestions regarding what PSR3 severity levels to use and icons that may be more intuitive.
    * Documentation ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) - __PASS__:  the changes do not require documentation.
    * Run it ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) - __ISSUE__: I logged into the jenkins build and played around with some relationship names/case roles, I found one way to break things otherwise it looked good, see list of workflows I tried and what broke below:
      * changing the label for the case manager role and then creating a case.  - PASS
      * changing names/labels of relationships types (successfully triggering the system status error :+1:) - PASS
      * creating a new relationship when creating a case type and saving in the default direction (saves and displays as expected). - PASS
      * creating a new relationship when creating a case type and saving in the OTHER (not the default) direction results in an API Error see screenshot below and more notes in the code review - **ISSUE**
Screenshot of Bug we found

* Developer standards
    * Technical impact ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech))- __PASS__: The change preserves compatibility with existing callers/code/downstream.
    * Code quality ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code))-  __ISSUE__: The changes are well documented with lots of helpful comments. The nature of this code is inherently confusing predating these changes there are names used as labels and a_bs that are b_as) We have some suggestions of how to improve readability of the code by renaming a property see code review for more details.
    * Test results ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) - __PASS__: The test results are all-clear.
    * Maintainability ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) -   __PASS__: The changes sufficiently improve test coverage AND add system checks that alert the user if they have a problematic relationship/case role set up :+1: see screenshot of new system status checks below.
New System Status Checks

**Comment by demeritcowboy** [comment]
Thanks - will take a look. You were probably hoping to never have to think about the a_b again so appreciate it.

File: CRM/Utils/Check/Component/Case.php [hunk 1]
```diff
@@ -162,4 +162,208 @@ public function checkNullTimestamps() {
...
+            5 => $dao->id2,
+            6 => htmlspecialchars($dao->nameab2),
+            7 => htmlspecialchars($dao->nameba2),
+          ]) .
+          '<br /><a href="' . CRM_Utils_System::docURL2('user/case-management/what-you-need-to-know#TODO', TRUE) . '">' .
+          ts('Read more about this warning') .
+          '</a>',
+        ts('Relationship Type Internal Name Duplicates'),
+        \Psr\Log\LogLevel::WARNING,
+        'fa-puzzle-piece'
```
[the remark this answers is not in the document]
Comment (demeritcowboy) [reply CRM/Utils/Check/Component/Case.php]: Done

File: CRM/Utils/Check/Component/Case.php [hunk 2]
```diff
@@ -162,4 +162,208 @@ public function checkNullTimestamps() {
...
+            3 => htmlspecialchars($dao->name_a_b),
+            4 => htmlspecialchars($dao->name_b_a),
+            5 => htmlspecialchars($dao->label_a_b),
+            6 => htmlspecialchars($dao->label_b_a),
+          ]) .
+          '<br /><a href="' . CRM_Utils_System::docURL2('user/case-management/what-you-need-to-know#TODO', TRUE) . '">' .
+          ts('Read more about this warning') .
+          '</a>',
+        ts('Relationship Type Ambiguity'),
+        \Psr\Log\LogLevel::WARNING,
```
[the remark this answers is not in the document]
Comment (demeritcowboy) [reply CRM/Utils/Check/Component/Case.php]: Added the similar check. Regarding `r-doc` I think the status checks should have some additional documentation for people, which is the user-guide proposed commit referenced in the description. I've updated the "read more" links in the status checks to point to where I think they would end up, but of course those links don't fully work live without the user-guide update.

File: CRM/Case/XMLProcessor/Process.php [hunk 3]
```diff
@@ -848,32 +842,67 @@ private function getBoolSetting($settingKey, $xmlTag, $default = 0) {
...
+   * status check.
+   * We do this to avoid requiring people to update their xml files which can
+   * be stored in external files we can't/don't want to edit.
    *
    * @param SimpleXMLElement $xml
    *
-   * @return string
+   * @return array[bool|string,string]
    */
   public function locateNameOrLabel($xml) {
```
[the remark this answers is not in the document]
Comment (demeritcowboy) [reply CRM/Case/XMLProcessor/Process.php]: Yes sorry I could have self-reviewed this part to help explain. The original idea of adding a new xml tag was based on something discussed during some rounds of cider, and I made the first PR towards it as a no-op that wouldn't take effect yet since the tag didn't exist yet, so it was still just doing what it did before. Then as I was clicking on more and more things I made a [comment](https://github.com/civicrm/civicrm-core/blob/d1a3ab5f2c73e6e48cb09ea8367302b2498d9e57/CRM/Case/BAO/CaseType.php#L188-L200) in one of the WIPs trying to convince myself ending with "hope this still makes sense to me later". Then a couple more clicks and reading it later, nope it didn't make sense, except for maybe one edge case. And then eventually I didn't need it all.

File: tests/phpunit/CRM/Case/XMLProcessor/ProcessTest.php [hunk 4]
```diff
@@ -240,30 +308,122 @@ protected function assertActivityAssignedToContactExists($assigneeContactId) {
...
+   *   identifier, it's an array key to use to look up something.
+   * @param $xmlString string
+   * @param $dontcare array We're re-using the data provider for two tests and
+   *   we don't care about those expected values.
+   * @param $expected array
    *
-   * @dataProvider xmlDataProvider
+   * @dataProvider xmlCaseRoleDataProvider
    */
-  public function testLocateNameOrLabel($xmlString, $expected) {
```
[the remark this answers is not in the document]
Comment (demeritcowboy) [reply tests/phpunit/CRM/Case/XMLProcessor/ProcessTest.php]: Ditto

File: ang/crmCaseType.js [hunk 5]
```diff
@@ -300,16 +308,25 @@
...
-        //  https://lab.civicrm.org/dev/core/issues/774
+        // TODO The `id` value is caught in between trying to support xml
+        // files where using the relType.id doesn't make it portable between
+        // installs and the xml doesn't have a concept of direction, but
+        // using name_a_b would conflict with it being used currently to set
+        // displaylabel elsewhere. So for now `id` is label.
         else {
           result.push({
+            machineName_a_b: relType.name_a_b,
+            machineName_b_a: relType.name_b_a,
```
[the remark this answers is not in the document]
Comment (demeritcowboy) [reply ang/crmCaseType.js]: Ok I see the point. I like xmlName. Not sure about what to call the other one.At the moment I've put `xmlNameOtherDirection`. `ClientName` to me means the client's name, like Steve. `UIName` seems like label. This also touches on some of the other thoughts I'm still collecting, and is the main reason I don't think this PR a perfect fix. The whole thing should be based around `civicrm_relationship_type.id` not names, and only converted to name going to/from xml where the original idea was portability between installs.

File: ang/crmCaseType.js [hunk 6]
```diff
@@ -460,31 +479,40 @@
...
+      } else {
+         // not a known relationship type, so create on-the-fly
+         CRM.loadForm(CRM.url('civicrm/admin/reltype', {action: 'add', reset: 1, label_a_b: roleName}))
+          .on('crmFormSuccess', function(e, data) {
+            var newType = _.values(data.relationshipType)[0];
+            roles.push({name: newType.name_b_a, displaylabel: newType.label_a_b});
+            // Assume that the case role should be A-B but add both directions as options.
+            $scope.relationshipTypeOptions.push({id: newType.name_a_b, text: newType.label_a_b});
+            if (newType.label_a_b != newType.label_b_a) {
+              $scope.relationshipTypeOptions.push({id: newType.name_b_a, text: newType.label_b_a});
```
[the remark this answers is not in the document]
Comment (demeritcowboy) [reply ang/crmCaseType.js]: Good catch for the add-on-the-fly-then-reverse-before-save. I've updated.

**Comment by demeritcowboy** [comment]
Jenkins test this please.

**Comment by demeritcowboy** [comment]
jenkins test this please

File: ang/crmCaseType.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (agh1) [reply ang/crmCaseType.js]: @demeritcowboy I had initially assumed that relationship type ID would be the best, but that would only really make sense if the whole case type were stored as relational tables rather than XML.  Right now, the case type only lives in XML.

For what to call the other one, how about `systemName` or just `name`?  The name, given from the perspective of the client, is the convention everywhere but in the XML.

File: ang/crmCaseType.js [hunk 7]
```diff
@@ -300,16 +308,25 @@
...
-        //  https://lab.civicrm.org/dev/core/issues/774
+        // TODO The `id` value is caught in between trying to support xml
+        // files where using the relType.id doesn't make it portable between
+        // installs and the xml doesn't have a concept of direction, but
+        // using name_a_b would conflict with it being used currently to set
+        // displaylabel elsewhere. So for now `id` is label.
         else {
           result.push({
+            xmlNameOtherDirection: relType.name_a_b,
+            xmlName: relType.name_b_a,
```
Comment (agh1) [inline ang/crmCaseType.js]: Ok I think these values got mixed up in a few places when re-labeling them.  I think these options are supposed to have the following pattern:

|   | `a_b` (the non-client's perspective, as in XML) | `b_a` (the client's perspective, as in the UI) |
| ----- | ----- | ----- |
| **`name`** | `xmlName` | `xmlNameOtherDirection` |
| **`label`** | `id` | `text` |

So that would mean the two new items here are reversed, so it should be

```js
            xmlName: relType.name_a_b,
            xmlNameOtherDirection: relType.name_b_a,
```

File: ang/crmCaseType.js [hunk 8]
```diff
@@ -300,16 +308,25 @@
...
             text: relType.label_b_a,
             id: relType.label_a_b
           });
 
           if (!isBidirectionalRelationship) {
             result.push({
+              // This looks wrong but at least for matching up to
+              // everything else xmlName needs to be backwards too.
+              xmlNameOtherDirection: relType.name_b_a,
+              xmlName: relType.name_a_b,
```
Comment (agh1) [inline ang/crmCaseType.js]: Similarly, this should be

```js
              xmlName: relType.name_b_a,
              xmlNameOtherDirection: relType.name_a_b,
```

Basically, the `xmlName` should always be the opposite direction of `text`, since the former is what is written to XML and the latter is what the user sees.

File: ang/crmCaseType.js [hunk 9]
```diff
@@ -354,9 +371,11 @@
 
       // go lookup and add client-perspective labels for $scope.caseType.definition.caseRoles
       _.each($scope.caseType.definition.caseRoles, function (set) {
-        _.each($scope.relationshipTypeOptions, function (relTypes) {
-          if (relTypes.text == set.name) {
+        _.each($scope.relationshipTypeOptionsAll, function (relTypes) {
+          if (relTypes.xmlName == set.name) {
             set.displaylabel = relTypes.id;
```
Comment (agh1) [inline ang/crmCaseType.js]: I think this is a spot where the original comparison was done backwards, but because it was just among labels, it didn't matter too much.  Now, it could be an issue.  Comparing the `xmlName` to the name stored in the XML is correct, but the displayLabel should be the `text`--the client's perspective, as used in the UI:

```suggestion
             set.displaylabel = relTypes.text;
```

File: ang/crmCaseType.js [hunk 10]
```diff
@@ -460,31 +479,55 @@
...
-              // Assume that the case role should be A-B but add both directions as options.
-              $scope.relationshipTypeOptions.push({id: newType.label_a_b, text: newType.label_a_b});
-              if (newType.label_a_b != newType.label_b_a) {
-                $scope.relationshipTypeOptions.push({id: newType.label_b_a, text: newType.label_b_a});
-              }
-              $scope.$digest();
-            });
+      if (matchingRole) {
+        if (!_.contains(names, matchingRole.xmlNameOtherDirection)) {
+          roles.push({name: matchingRole.xmlNameOtherDirection, displaylabel: matchingRole.text});
```
Comment (agh1) [inline ang/crmCaseType.js]: The `name` here (and in the lookup above) should be `matchingRole.xmlName`, since it's the value that would be going in the XML (from the non-client's perspective).  The `displaylabel` is correct here, since `matchingRole.text` will be what the user is seeing (from the client's perspective).

File: ang/crmCaseType.js [hunk 11]
```diff
@@ -460,31 +479,55 @@
...
+          .on('crmFormSuccess', function(e, data) {
+            var newType = _.values(data.relationshipType)[0];
+            roles.push({name: newType.name_b_a, displaylabel: newType.label_a_b});
+            // Assume that the case role should be A-B but add both directions as options.
+            var newRelTypeOption = {
+              xmlNameOtherDirection: newType.name_b_a,
+              xmlName: newType.name_a_b,
+              id: newType.name_a_b,
+              text: newType.label_a_b
+            };
```
Comment (agh1) [inline ang/crmCaseType.js]: This got the directions and the names a little mixed up.  The `xmlName` should be the same as `roles.name` above.  The `text` is correctly the same as `roles.displaylabel` above.  `id` should be the label counterpart to `xmlName`, and `xmlNameOtherDirection` should be the name counterpart to `text`.  The result should look like this:

```js
            var newRelTypeOption = {
              xmlName: newType.name_b_a,
              xmlNameOtherDirection: newType.name_a_b,
              id: newType.label_b_a,
              text: newType.label_a_b
            };
```

File: ang/crmCaseType.js [hunk 12]
```diff
@@ -460,31 +479,55 @@
...
+            $scope.relationshipTypeOptions.push(newRelTypeOption);
+            $scope.relationshipTypeOptionsAll.push(newRelTypeOption);
+            // Add the other direction if different.
+            if (newType.label_a_b != newType.label_b_a) {
+              newRelTypeOption = {
+                xmlNameOtherDirection: newType.name_a_b,
+                xmlName: newType.name_b_a,
+                id: newType.name_b_a,
+                text: newType.label_b_a
+              };
```
Comment (agh1) [inline ang/crmCaseType.js]: Same issues as above, so it should be:

```js
              newRelTypeOption = {
                xmlName: newType.name_a_b,
                xmlNameOtherDirection: newType.name_b_a,
                id: newType.label_a_b,
                text: newType.label_b_a
              };
```

**Review by agh1 — CHANGES_REQUESTED** [review]
Just a handful of A and B, label and name issues.  Hopefully the table below is helpful.

**Comment by demeritcowboy** [comment]
@agh1 I was just looking at that. The on-the-fly push seemed backwards. I'm also going to look at wheher creating and pushing these new objects around is even necessary. It seems like relType from the api call could almost be used in-place (except maybe id).

**Comment by colemanw** [comment]
@demeritcowboy @alifrumin @agh1 thank you all for your efforts on this!

Just a general note incase it is relevant: api4 has now been merged to core and as a general rule we are not adding any legacy support to api4 (at least not up-front for "pre existing conditions"). So when Case and CaseType apis are added they should probably support name only.

**Comment by demeritcowboy** [comment]
Ok. What I meant by the api reference is there's 4 similar objects being floated around these case type edit screens and they all have different member variables and one is backwards but they are all referencing the same underlying data - the data returned at the beginning by the api call to retrieve relationship types. What gets sent back to the server would stay the same I just want to take look-see whether some of the a_b/b_a confusion can be reduced somehow while inside these edit screens. Either by not relabelling variables or maybe by more relabelling of variables. Code cleanup wasn't really my goal here but I'm kind of all-in on this one already.

**Comment by agh1** [comment]
@demeritcowboy assuming all of those changes I proposed work, I think at the very least `xmlNameOtherDirection` could be dropped.  The fact that we can't think of a good name for it may be a sign that it doesn't really correspond to anything.  The other items do have a use:

- `xmlName`: the way a role is recorded in the XML
- `text`: the label that you see when you're picking roles
- `id`: the value that comes back from the "Add role" drop-down (but only because that's what set as the `id` in `relationshipTypeOptions`)

I *think* that it might be possible to ditch what is now the `id` and instead send the XML-oriented name values (what we're storing as `xmlName`) as `id`.  The drop-down is only going to show what we send it, so as long as we're consistent on both sending and receiving, that should be fine.  Assuming that works, all of the places that use `xmlName` could begin using the new `id`, and that would mean that the relationshipTypeOptions could have just `id` and `text`.

In describing this, I also realized that the case role table will always have trouble if the XML has a label that doesn't match the corresponding name from the relationship type.  There's no equivalent to `CRM_Case_XMLProcessor_Process::locateNameOrLabel()` over in Angularland.

For example, let's say there's a relationship type with the following values:
- `name_a_b`: Grandma is
- `label_a_b`: Grandchild of
- `name_b_a`: Grandma of
- `label_b_a`: Grandparent of

Let's say that the client is a child and the grandparent is the other role.  Upon upgrade, the XML will say that the role name is `Grandparent of` (the label from the perspective of the non-client), since that has been the convention so far.  This will be looking for that among the `xmlName` values, and it won't be there.  (There will be a `Grandma is` and a `Grandma of`, the latter being what we'd ultimately want, but it has no way of knowing that.)

Assuming that we go down the direction of a slimmer `relationshipTypeOptions`, we could just assemble an object that keys the names to their corresponding labels (in the same direction).  If the comparison at 
https://github.com/civicrm/civicrm-core/commit/86e0cd5a1b1f280f0360727a0bf6a503132db7fb#diff-f42508faf29afc0f8ecd02bcb3d6be66R374-R375
fails, we could see if the value is a label, get the corresponding name, and try looking for that.

**Comment by demeritcowboy** [comment]
Probably yes to the first part. To the second part it's dealt with elsewhere in the PR. The case type defn that you get when you retrieve a case type has the name in it regardless of what the actual xml has.

File: ang/crmCaseType.js (hunk 12 above)
Comment (agh1) [reply ang/crmCaseType.js]: I also realize that we should probably push to `$scope.defaultRelationshipTypeOptions` since there's no way to say that your new role should be default.

**Comment by agh1** [comment]
@demeritcowboy fabulous--you're right.  I forgot that you process it on its way out from the XML.

File: ang/crmCaseType.js (hunk 12 above)
Comment (demeritcowboy) [reply ang/crmCaseType.js]: My testing notes say this was pre-existing so I ignored it. But yes that seems doable.

**Comment by demeritcowboy** [comment]
See https://github.com/civicrm/civicrm-core/pull/15483 for version which is the same server-side but includes code cleanup in the case type edit screens as mentioned above.

File: ang/crmCaseType.js (hunk 12 above)
Comment (demeritcowboy) [reply ang/crmCaseType.js]: >  The fact that we can't think of a good name for it may be a sign that it doesn't really correspond to anything.

This turns out to be true! In the new PR there's just xmlName, and then id and text stay because select2 is expecting variables with those names.

> I also realize that we should probably push to `$scope.defaultRelationshipTypeOptions` since there's no way to say that your new role should be default.

In the new PR this ends up happening almost automatically, because that object gets combined.
